tailor_image/create_image.py

```diff
@@ -141,41 +146,79 @@ def create_image(name: str, distribution: str, apt_repo: str, release_track: str
 
     run_command(command, env=env, cwd='/tmp')
 
-    # TODO(gservin): If we build more that one bare metal image at the same time, we can have a race condition here
     if build_type == 'bare_metal' and publish:
-        update_image_index(release_track, apt_repo, common_config, image_name)
-
-
-def update_image_index(release_track, apt_repo, common_config, image_name):
+        update_image_index(release_track, apt_repo, common_config, image_name, set_default)
+
+
+def update_image_index(release_track, apt_repo, common_config, image_name, default_image=False):
+    """Updates the index file used to track bare metal images
+
+    Current format:
+    {
+      "<release_label>": {
+        "<flavour>": {
+          "version": "<organization>_<flavour>_<distribution>_<release_label>_<date>",
+          "checksums": {
+            "<organization>_<flavour>_<distribution>_<release_label>_<date>": "<md5sum_of_image>",
+          }
+        },
+        "<flavour_2>": {
+        }
+      }
+    }
+    """
     index_key = release_track + '/images/index'
-    s3_object = boto3.resource("s3").Bucket(apt_repo)
-    json.load_s3 = lambda f: json.load(s3_object.Object(key=f).get()["Body"])
-    json.dump_s3 = lambda obj, f: s3_object.Object(key=f).put(Body=json.dumps(obj, indent=2))
-    _, _, distribution, release_label, _ = image_name.split('_')
+    s3_bucket = boto3.resource("s3").Bucket(apt_repo)
+
+    # Wait until no lock file exists to avoid race condition
+    lock = s3_bucket.Object(f'{index_key}.lock')
```

Comment:
https://stackoverflow.com/questions/33347899/locking-an-s3-object-best-practice

This is better, but probably still not optimal. I imagine we eventually want something like tailor-distro that uploads all the images and writes the index in one shot.

Reply:
Thanks! I read about the SQS but wasn't sure how complicated was going to be. I've changed it now to use tags as recommended on the link